### PR TITLE
Adding Get/Set String to IDistributedCache

### PR DESCRIPTION
**Pull request by BrennanConroy**
#169 
@JunTaoLuo @Tratcher @Eilon @muratg

### CODE REVIEW

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs [hunk 1]
```diff
@@ -43,5 +44,35 @@ public static Task SetAsync(this IDistributedCache cache, string key, byte[] val
 
             return cache.SetAsync(key, value, new DistributedCacheEntryOptions());
         }
+
+        public static void SetString(this IDistributedCache cache, string key, string value)
+        {
+            cache.Set(key, Encoding.UTF8.GetBytes(value));
```
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: What happens if `value` is `null` ?

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs [hunk 2]
```diff
@@ -43,5 +44,35 @@ public static Task SetAsync(this IDistributedCache cache, string key, byte[] val
 
             return cache.SetAsync(key, value, new DistributedCacheEntryOptions());
         }
+
+        public static void SetString(this IDistributedCache cache, string key, string value)
```
Comment (JunTaoLuo) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: Doc comments?

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs (hunk 1 above)
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: From the doc it throws, it needs to be handled then.

https://msdn.microsoft.com/en-us/library/ds4kkd55(v=vs.110).aspx

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs (hunk 2 above)
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: You need overloads that take DistributedCacheEntryOptions.

**Comment by BrennanConroy** [comment]
🆙 📅

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs [hunk 3]
```diff
@@ -43,5 +44,93 @@ public static Task SetAsync(this IDistributedCache cache, string key, byte[] val
...
+        /// Asynchronously sets a string in the specified cache with the specified key.
+        /// </summary>
+        /// <param name="cache">The cache in which to store the data.</param>
+        /// <param name="key">The key to store the data in.</param>
+        /// <param name="value">The data to store in the cache.</param>
+        /// <param name="options">The cache options for the entry.</param>
+        /// <returns>A task that represents the asynchronous set operation.</returns>
+        public static Task SetStringAsync(this IDistributedCache cache, string key, string value, DistributedCacheEntryOptions options)
+        {
+            if (value == null)
```
Comment (sebastienros) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: It should be discussed, but I think we should handle `null` values in a way that a user could really mean "store a null" value, and get a "null" value back. It's a valid value for a string and can have meaning. Even if it means "delete the entry".

And it should not conflict with empty strings either, which are another valid value.

Comment (Eilon) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: Not sure about the "delete the entry" behavior. But I would think we should persist/retrieve `null` strings as a valid value.

Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: There is no ContainsKey, so storying a null entry is indistinguishable from ignoring/removing the entry. We can store string.Empty as byte[0], but we don't have a meaningful way to serialize and store null.

Comment (Eilon) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: We can put some token value to represent `null`. We can store things however we like...

**Comment by BrennanConroy** [comment]
Updated

File: test/Microsoft.Extensions.Caching.Memory.Tests/MemoryCacheSetAndRemoveTests.cs [hunk 4]
```diff
@@ -84,6 +84,16 @@ public void SetAlwaysOverwrites()
         }
 
         [Fact]
+        public void SetNullValueThrows()
```
Comment (sebastienros) [inline test/Microsoft.Extensions.Caching.Memory.Tests/MemoryCacheSetAndRemoveTests.cs]: So we currently can't add a null value to a memory cache, even today?

Comment (Eilon) [inline test/Microsoft.Extensions.Caching.Memory.Tests/MemoryCacheSetAndRemoveTests.cs]: Yeah memory cache should still allow this. Just distributed should block. No?

File: src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs [hunk 5]
```diff
@@ -34,14 +40,115 @@ public static void Set(this IDistributedCache cache, string key, byte[] value)
...
+        /// <summary>
+        /// Sets a string in the specified cache with the specified key.
+        /// </summary>
+        /// <param name="cache">The cache in which to store the data.</param>
+        /// <param name="key">The key to store the data in.</param>
+        /// <param name="value">The data to store in the cache.</param>
+        /// <exception cref="System.ArgumentNullException">Thrown when <paramref name="key"/> or <paramref name="value"/> is null.</exception>
+        public static void SetString(this IDistributedCache cache, string key, string value)
+        {
+            if (value == null)
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Abstractions/DistributedCacheExtensions.cs]: It's strange to just check the value here. I assume you depend on the called function to check the key? If so, you can do the same for value.

**Comment by BrennanConroy** [comment]
🆙 📅

**Comment by Tratcher** [comment]
:shipit:

**Pull request merged**